### ISSUE TITLE
Async timeout triggers very late

### ISSUE THREAD

**Issue by asos-pavlospavlopoulos** [StackExchange/StackExchange.Redis#1092]
We are seeing very late async timeouts in some cases:

```
Timeout awaiting response (30469ms elapsed, timeout is 3000ms), inst: 0, qs: 92, in: 0, serverEndpoint: Unspecified/example.redis.cache.windows.net:6380, mgr: 10 of 10 available, clientName: Web.Api_IN_0, IOCP: (Busy=0,Free=1000,Min=256,Max=1000), WORKER: (Busy=303,Free=32464,Min=256,Max=32767), v: 2.0.519.65453 (Please take a look at this article for some common client-side issues that can cause timeouts: https://stackexchange.github.io/StackExchange.Redis/Timeouts)
```

This is just an example, we've seen a max of up to 60 seconds with a configured timeout of 3sec. It is also worth noting that there seems to be 10/10 available threads in the internal thread pool but the system worker thread count has exceeded the minimum. We have checked the machine resources at the time of the timeouts and everything is normal from a CPU and memory perspective.

Any idea what could be the cause of such a big drift in actual timeout and this thread behaviour?

**Comment by mgravell** [comment]
Async timeouts are processed by a timer that depends on the thread-pool. As
it happens, 519 has a known sync-over-async problem which can lead to
thread-pool exhaustion, which could prevent the timer from firing. As such,
it is possible that the 2.0.560+ release will resolve this. This is
available on nuget, but is not currently advertised while we dogfood it.

**Comment by mgravell** [comment]
We are aware of a problem here (relating to a race condition in the write-lock); we're dogfooding a fix locally, and hope to deploy an update in the next day or two.

**Comment by mgravell** [comment]
The combined improvements in 2.0.588 and 2.0.571 should, we believe, address a wide range of timeout and performance problems. I'm closing all timeout issues as "believed fixed in 2.0.588"; if you still see problems, please report them as new issues so we can start investigation with fresh eyes.
